content/blog/2023-12-01-task-lifecycle.markdown

```diff
@@ -0,0 +1,470 @@
+	cpu_startup_entry(CPUHP_ONLINE);
+}
+
+```
+
+After this, it basically just enters the scheduler. We're now almost fully task
+driven, and our journey picks back up inside of the init task.
+
+> Hint: you can see where the `init_task` vs the init task is used in the
+> console. The `init_task` has PID 0, so appears as `T0` in the log line
+> metadata. The init task has PID 1, so appears as `T1`.
+
+
+## The init task
+
+When we created the init task, we set the entry point to be the `kernel_init()`
+function. Execution simply begins from here once we get scheduled. The very
+first thing we do is wait for the kthreadd task to be up and running: bad things
+would happen if we tried to use kernel services that rely on kthreads otherwise.
+
+The wait mechanism itself also interacts with the scheduler a lot: starting with
+a common `struct completion` object, the waiting task registers itself as
+awaiting the object to complete. Specifically, it adds its task handle to a
+queue on the completion object. It then loops calling `schedule()`, which yields
+itself to the scheduler, until the completion object is flagged as done.
+Somewhere else, such as another task, the completion object will eventually be
+marked as completed. As part of this the task marking the completion tries to
+wake up a task that might have registered itself as waiting earlier.
+
+In the kthreadd case, the init task waits on a completion object that
+`kernel_init()` marks completed after creating kthreadd. We could technically
+avoid this synchronization altogether just by creating kthreadd first, but then
+the init task wouldn't have PID 1.
+
+The rest of the init task wraps up the initialisation stage as a whole. Mostly
+it moves the system into the 'running' state after freeing any memory marked as
+for initialisation only (set by `__init` annotations).
+
+Once fully initialised and running, the init tasks attempts to execute the
+userspace init program.
+
+```c
+	if (execute_command) {
+		ret = run_init_process(execute_command);
+		if (!ret)
+			return 0;
+		panic("Requested init %s failed (error %d).",
+		      execute_command, ret);
+	}
+
+	if (CONFIG_DEFAULT_INIT[0] != '\0') {
+		ret = run_init_process(CONFIG_DEFAULT_INIT);
+		if (ret)
+			pr_err("Default init %s failed (error %d)\n",
+			       CONFIG_DEFAULT_INIT, ret);
+		else
+			return 0;
+	}
+
+	if (!try_to_run_init_process("/sbin/init") ||
+	    !try_to_run_init_process("/etc/init") ||
+	    !try_to_run_init_process("/bin/init") ||
+	    !try_to_run_init_process("/bin/sh"))
+		return 0;
+
+	panic("No working init found.  Try passing init= option to kernel. "
+	      "See Linux Documentation/admin-guide/init.rst for guidance.");
+```
+
+There are several locations it will attempt, in order:
+
+1. Ramdisk file set by `rdinit=` boot command line parameter, with default path
+   `/init`.
+2. File set by `init=` boot command line parameter.
+3. `/sbin/init`
+4. `/etc/init`
+5. `/bin/init`
+6. `/bin/sh`
```

Comment:
That said, specifying that the init script is to be found on the filesystem should accomplish the above clarification.